internal/api/v1/application/importgit.go

```diff
@@ -94,3 +146,81 @@ func (hc Controller) ImportGit(c *gin.Context) apierror.APIErrors {
 	})
 	return nil
 }
+
+func getRepository(ctx context.Context, log logr.Logger, gitRepo, url, revision string) error {
+	if revision == "" {
+		// Input A: repository, no revision.
+		log.Info("importgit, cloning simple", "url", url)
+		_, err := generalClone(ctx, gitRepo, url)
+		return err
+	}
+
+	// Input B or C: Attempt to treat as B (revision is branch name)
+
+	log.Info("importgit, cloning branch", "url", url, "revision", revision)
+	_, err := branchClone(ctx, gitRepo, url, revision)
+	if err == nil {
+		// Was branch name, done.
+		return nil
+	}
+	if !strings.Contains(err.Error(), `couldn't find remote ref`) || !plumbing.IsHash(revision) {
+		// Some other error, or the revision does not look like a commit id (C)
+
+		return err
+	}
+
+	// Attempt input C: revision might be commit id.
+	// 2 stage process - A simple clone followed by a checkout
+
+	log.Info("importgit, cloning simple, commit id", "url", url)
+	repository, err := generalClone(ctx, gitRepo, url)
+	if err != nil {
+		return err
+	}
+
+	log.Info("importgit, resolve", "revision", revision)
+	hash, err := repository.ResolveRevision(plumbing.Revision(revision))
+	if err != nil {
+		return err
+	}
+
+	log.Info("importgit, resolved", "revision", hash)
+
+	checkout, err := repository.Worktree()
+	if err != nil {
+		return err
+	}
+
+	log.Info("importgit, checking out", "url", url, "revision", hash)
+
+	return checkout.Checkout(&git.CheckoutOptions{
+		Hash:  *hash,
+		Force: true,
+	})
+}
+
+func branchClone(ctx context.Context, gitRepo, url, revision string) (*git.Repository, error) {
+	repository, err := git.PlainCloneContext(ctx, gitRepo, false, &git.CloneOptions{
+		URL:           url,
+		SingleBranch:  true,
+		ReferenceName: plumbing.NewBranchReferenceName(revision),
+		Depth:         1,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return repository, nil
+}
+
+func generalClone(ctx context.Context, gitRepo, url string) (*git.Repository, error) {
+	repository, err := git.PlainCloneContext(ctx, gitRepo, false, &git.CloneOptions{
+		URL:   url,
+		Depth: 1,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return repository, nil
+}
```

Comment:
It seems that the only difference are the CloneOptions, and we are not adding more info to the error, so we can just

```go
func branchClone(ctx context.Context, gitRepo, url, revision string) (*git.Repository, error) {
	opts := &git.CloneOptions{
		URL:           url,
		Depth:         1,
		SingleBranch:  true,
		ReferenceName: plumbing.NewBranchReferenceName(revision),
	}
	return git.PlainCloneContext(ctx, gitRepo, false, opts)
}

func generalClone(ctx context.Context, gitRepo, url string) (*git.Repository, error) {
	opts := &git.CloneOptions{
		URL:           url,
		Depth:         1,
	}
	return git.PlainCloneContext(ctx, gitRepo, false, opts)
}
```

Reply:
Yes. Missed that.

Reply:
Done, commit 91d2efd273b228842207c707fafd5752303327b6